selfdrive/controls/radard.py

```diff
@@ -204,41 +212,44 @@ def __init__(self, radar_ts: float, delay: int = 0):
 
     self.ready = False
 
-  def update(self, sm: messaging.SubMaster, rr: Optional[car.RadarData]):
+  def update(self, sm: messaging.SubMaster, radar_data: car.RadarData):
     self.current_time = 1e-9*max(sm.logMonoTime.values())
 
-    radar_points = []
-    radar_errors = []
-    if rr is not None:
-      radar_points = rr.points
-      radar_errors = rr.errors
+    radar_points = radar_data.points
+    radar_errors = radar_data.errors
 
     if sm.updated['carState']:
       self.v_ego = sm['carState'].vEgo
       self.v_ego_hist.append(self.v_ego)
     if sm.updated['modelV2']:
       self.ready = True
 
-    ar_pts = {}
-    for pt in radar_points:
-      ar_pts[pt.trackId] = [pt.dRel, pt.yRel, pt.vRel, pt.measured]
+    if radar_data.parseCompleted:
+      ar_pts = {}
+      for pt in radar_points:
+        ar_pts[pt.trackId] = [pt.dRel, pt.yRel, pt.vRel, pt.measured]
 
-    # *** remove missing points from meta data ***
-    for ids in list(self.tracks.keys()):
-      if ids not in ar_pts:
-        self.tracks.pop(ids, None)
+      # *** remove missing points from meta data ***
+      for ids in list(self.tracks.keys()):
+        if ids not in ar_pts:
+          self.tracks.pop(ids, None)
 
-    # *** compute the tracks ***
-    for ids in ar_pts:
-      rpt = ar_pts[ids]
+      # *** compute the tracks ***
+      for ids in ar_pts:
+        rpt = ar_pts[ids]
 
-      # align v_ego by a fixed time to align it with the radar measurement
-      v_lead = rpt[2] + self.v_ego_hist[0]
+        # align v_ego by a fixed time to align it with the radar measurement
+        v_lead = rpt[2] + self.v_ego_hist[0]
 
-      # create the track if it doesn't exist or it's a new track
-      if ids not in self.tracks:
-        self.tracks[ids] = Track(v_lead, self.kalman_params)
-      self.tracks[ids].update(rpt[0], rpt[1], rpt[2], v_lead, rpt[3])
+        # create the track if it doesn't exist or it's a new track
+        if ids not in self.tracks:
+          self.tracks[ids] = Track(ids, v_lead, self.kalman_params)
+        self.tracks[ids].update(rpt[0], rpt[1], rpt[2], v_lead, rpt[3])
+    else:
+      # *** no radar points, keep existing tracks, update v_lead
+      for track in self.tracks.values():
+        v_lead = track.vRel + self.v_ego_hist[0]
+        track.update_vlead(v_lead)
```

Comment:
does this preserve behavior or change it? Previously when returning None, it just skipped running any updates until it got the next header message. Now it runs `update_vlead` while there's no header message

Reply:
previously the tracks were always removed/added because it run only when radarData was not None. Now it's possible that update gets executed before RI seen trigger message, so in that case it only updates vLead based on last vLead value and new vEgo

Reply:
Right, but it would only publish that update once, never extrapolating the vLead data (correct?). Now we have two choices: publish the tracks with the same data, except updating vLead, or publish the tracks with the same data without modification (until we have an update).

Reply:
Though isn't it intended we only do this once, since we expect delay in the radar measurement? This is only executed for Honda which sets `self.delay` in RadarInterface.

Reply:
Not sure if i follow that hondas delay actually mean.

```
   modelV2        modelV2         modelV2

----- | ------------ | ------------- |
--- | ---------------- | ---------------- |
 trigger            trigger            trigger
                       
```
^This is the case that this covers. If radar is slow (some of them run at sub 20hz), the update may be called without any points from RadarInterface (second modelV2 in the graph above), since its tied to modelV2 now. carState updates at 100Hz, so we get new vEgo during that time, so we update existing lead speeds accordingly.